kafka/mockcluster.go

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka
+
+import "C"
+import "unsafe"
+
+/*
+#include <stdlib.h>
+#include "select_rdkafka.h"
+#include "glue_rdkafka.h"
+*/
+import "C"
+
+type MockCluster struct {
+	rk       *C.rd_kafka_t
+	mcluster *C.struct_rd_kafka_mock_cluster_s
+	cConf    *C.rd_kafka_conf_t
+}
+
+/* NewMockCluster provides a mock Kafka cluster with a configurable
+ * number of brokers that support a reasonable subset of Kafka protocol
+ * operations, error injection, etc.
+ *
+ * Mock clusters provide localhost listeners that can be used as the bootstrap
+ * servers by multiple rd_kafka_t instances.
+ *
+ * Currently supported functionality:
+ * - Producer
+ * - Idempotent Producer
+ * - Transactional Producer
+ * - Low-level consumer
+ * - High-level balanced consumer groups with offset commits
+ * - Topic Metadata and auto creation
+ *
+ * @remark This is an experimental public API that is NOT covered by the
+ *         librdkafka API or ABI stability guarantees.
+ *
+ * @warning THIS IS AN EXPERIMENTAL API, SUBJECT TO CHANGE OR REMOVAL.
+ */
+func NewMockCluster(brokerCount int) (*MockCluster, error) {
+
+	mc := &MockCluster{}
+
+	cErrstr := (*C.char)(C.malloc(C.size_t(256)))
+	defer C.free(unsafe.Pointer(cErrstr))
+
+	mc.cConf = C.rd_kafka_conf_new()
+
+	mc.rk = C.rd_kafka_new(C.RD_KAFKA_PRODUCER, mc.cConf, cErrstr, 256)
```

Comment:
There is an issue here when `Close()` is called which causes a crash.

In the documentation for [rd_kafka_new](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#a63d5cd86ab1f77772b2be170e1c09c24) it says:

> conf is an optional struct created with [rd_kafka_conf_new()](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#aa7459bd22e8cfa81aa8c2480a4a0304c) that will be used instead of the default configuration. The conf object is freed by this function on success and must not be used or destroyed by the application sub-sequently. See [rd_kafka_conf_set()](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#abb1b319278333e8cdee9442da7f135e8) et.al for more information.

This means that on success there should be no calls to `C.rd_kafka_conf_destroy(mc.cConf)` which is done in `Close()`

Reply:
Ok. I removed the call to the conf_destroy function to prevent problems ;-) Thanks